content/blog/2020-12-18-type-and-effect-system.md

```diff
@@ -0,0 +1,93 @@
+
+Below is a sample program demonstrating functions and their properties:
+
+```
+add5 : int -> int = add 5;
+twice : (int -> int) -> int -> int = \f : int -> int. \x : int. f (f x);
+print_thrice : (int -> unit !output) = \x : int. {
+  _ : unit !output = print x;
+  _ : unit !output = print x;
+  _ : unit !output = print x;
+}
+_ : unit !output = print_thrice (twice add5 2);
+```
+
+# Implementation
+The language is written entirely in OCaml, using Jane Street Core as standard library, ocamllex as lexer-generator, and Menhir as parser-generator. The provided implementation includes a typechecker, an interpreter, and a sample optimization pass. The `Or_error` monad and its corresponding `let%bind`/`let%map` monadic syntax is used throughout the implementation to capture and thread through errors.
+
+## Typechecker
+The typechecker takes a parsed program AST as input and checks the types. For example, the `Assign (x, t, effs, e)` rule determines the type of the right-hand-side expression and checks if it has type `t` and effects `e` before adding `x : t` to the type context. The `If (e, s1, s2)` rule ensures that `e` has type `bool`, and takes the intersection of the type contexts from `s1` and `s2`. Since all expressions must be bound by an assignment, and only `Apply` expressions can cause side effects, all effects must be captured by assignment statements. 
+
+The standard library provides a mapping from function names to function types, which is used as the starting type context for typechecking. This had the unfortunate constraint of disallowing `print` from supporting a variable number of arguments; varargs support would require an extra wrapper around types in the type context or a new `Any` type, neither of which were appealing options for this project.
+
+## Interpreter
+The interpreter takes a parsed program AST as input and executes the program, relying on the typechecker as a preprocessing step to better enforce correctness. Starting from an empty map from variable names to their values, each statement of the list of top-level statements comprising the entire program is executed in order, with the resulting context being threaded from one statement to the next. 
+
+The most difficult part of this module was correctly evaluating lambdas and closing around the variables needed, especially in partial application situations. To that end, `Lambda (x, t, s)` had to be extended in the code to `Lambda (x, t, ctx, s)`, with `ctx` representing the context of the closure upon creation. Then, partial application could be implemented as nested lambda expressions, and evaluating one layer of a lambda stack would encode its argument into the context of the next layer. 
```

Comment:
Sounds about right. Just to be clear here, though, there wasn't anything particularly effecty about this, right? It's how you would do it if you were implementing an interpreter for a language without effects too?

Reply:
yep! added a bit in reference to the extensions section that talks about effects at runtime